cmd/cosign/cli/copy/copy.go

```diff
@@ -48,41 +50,49 @@ func CopyCmd(ctx context.Context, regOpts options.RegistryOptions, srcImg, dstIm
 	dstRepoRef := dstRef.Context()
 
 	remoteOpts := regOpts.GetRegistryClientOpts(ctx)
+
+	pusher, err := remote.NewPusher(remoteOpts...)
+	if err != nil {
+		return err
+	}
+
+	g, gctx := errgroup.WithContext(ctx)
+	g.SetLimit(runtime.GOMAXPROCS(0))
+
 	root, err := ociremote.SignedEntity(srcRef, ociremote.WithRemoteOptions(remoteOpts...))
 	if err != nil {
 		return err
 	}
 
-	if err := walk.SignedEntity(ctx, root, func(ctx context.Context, se oci.SignedEntity) error {
+	if err := walk.SignedEntity(gctx, root, func(ctx context.Context, se oci.SignedEntity) error {
 		// Both of the SignedEntity types implement Digest()
 		h, err := se.(interface{ Digest() (v1.Hash, error) }).Digest()
 		if err != nil {
 			return err
 		}
 		srcDigest := srcRepoRef.Digest(h.String())
 
-		// Copy signatures.
-		if err := copyTagImage(ociremote.SignatureTag, srcDigest, dstRepoRef, force, remoteOpts...); err != nil {
-			return err
-		}
-		if sigOnly {
-			return nil
-		}
+		for i, tm := range []tagMap{ociremote.SignatureTag, ociremote.AttestationTag, ociremote.SBOMTag} {
+			src, err := tm(srcDigest, ociremote.WithRemoteOptions(remoteOpts...))
+			if err != nil {
+				return err
+			}
 
-		// Copy attestations
-		if err := copyTagImage(ociremote.AttestationTag, srcDigest, dstRepoRef, force, remoteOpts...); err != nil {
-			return err
-		}
+			dst := dstRepoRef.Tag(src.Identifier())
+			g.Go(func() error {
+				return copyImage(ctx, pusher, src, dst, force, remoteOpts...)
+			})
 
-		// Copy SBOMs
-		if err := copyTagImage(ociremote.SBOMTag, srcDigest, dstRepoRef, force, remoteOpts...); err != nil {
-			return err
+			if i == 0 && sigOnly {
```

Comment:
Nit: I don't love that this depends on the order in the `tagMap`.

My preference might be to have a small helper function like:

```go
func whichTagsToCopyFromArgs(sigOnly bool) []tagMap {
  if sigOnly {
    return []tagMap{ociremote.SignatureTag}
  }
  return []tagMap{ociremote.SignatureTag, ociremote.AttestationTag, ociremote.SBOMTag}
}
```

This is extremely forwards-compatible with https://github.com/sigstore/cosign/issues/2002 and IMO properly separates parsing args from "what does this command do?"

Reply:
Not quite as simple as your helper (we avoid pushing the entity itself as well), but added something like that.